### PR TITLE
    [Proposal] Breaking change `zero-shot-object-detection` for improved     consistency.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

This is a proposal to modify the output of `zero-shot-object-detection`
to provide better alignment with other pipelines.

The output is now strictly the same as `object-detection` whereas before
it would output lists of lists.

The name `candidate_labels` is used throughout for consistency with
other `zero-shot` pipelines.

The pipeline is changed to `ChunkPipeline` to support batching cleanly.

This removes all the lists and list of lists shenanigans, it's now a
matter of the base pipeline handling all this not this specific one.

### **Breaking change**
It did remove complex calls potentials `pipe(images = [image1, image2],
text_queries=[candidates1, candidates2])` to support only
`pipe([{"image": image1, "candidate_labels": candidates1}, {"image": image2, "candidate_labels": candidates2}])`
when dealing with lists and/or datasets.
We could keep them, but it will add a lot of complexity to the code
base, since the pipeline is rather young, I'd rather break to keep the
code simpler, but we can revert this.

### **Breaking change**
The name of the argument is now `image` instead of
`images` since it expects by default only 1 image. This is revertable
like the previous one.

### **Breaking change**
The types is now simplified and flattened:
`pipe(inputs) == [{**object1}, {**object2}]`
instead of the previous
`pipe(inputs) == [[{**object1}, {**object1}], [{**object2}]]`
Where the different instances would be grouped by candidate labels
within lists.
IMHO this is not really desirable, since it would output empty lists and
is only adding superflous indirection compared to
`object-detection`.

It is relatively change free, meaning the results are the same for large models.
It does change computation however since now the batching is handled by the pipeline
itself. It **did** change the results for the small models so there
seems to be a real difference in how the models handles this. Since it didn't affect any of the large tests I think it's acceptable.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/pipelines/zero_shot_object_detection.py [hunk 1]
```diff
@@ -28,11 +20,33 @@
...
+
+    ```python
+    >>> from transformers import pipeline
+
+    >>> detector = pipeline(model="google/owlvit-base-patch32", task="zero-shot-object-detection")
+    >>> detector(
+    ...     "http://images.cocodataset.org/val2017/000000039769.jpg",
+    ...     candidate_labels=["cat", "couch"],
+    ... )
+    [{'score': 0.287, 'label': 'cat', 'box': {'xmin': 324, 'ymin': 20, 'xmax': 640, 'ymax': 373}}, {'score': 0.254, 'label': 'cat', 'box': {'xmin': 1, 'ymin': 55, 'xmax': 315, 'ymax': 472}}, {'score': 0.121, 'label': 'couch', 'box': {'xmin': 4, 'ymin': 0, 'xmax': 642, 'ymax': 476}}]
```
Comment (sgugger) [inline src/transformers/pipelines/zero_shot_object_detection.py:38]: You can use new lines instead of space to make the result more readable.

File: src/transformers/pipelines/zero_shot_object_detection.py [hunk 2]
```diff
@@ -119,49 +118,59 @@ def _sanitize_parameters(self, **kwargs):
...
-        target_sizes = [torch.IntTensor([[img.height, img.width]]) for img in images]
-        target_sizes = torch.cat(target_sizes)
-        inputs = self._processor(text=inputs["text_queries"], images=images, return_tensors="pt")
-        return {"target_sizes": target_sizes, "text_queries": text_queries, **inputs}
+        try:
+            image = load_image(inputs["image"])
+        except Exception:
+            import ipdb
+
+            ipdb.set_trace()
```
Comment (sgugger) [inline src/transformers/pipelines/zero_shot_object_detection.py:126]: Should this be cleaned up?

File: src/transformers/pipelines/zero_shot_object_detection.py [hunk 3]
```diff
@@ -119,49 +118,59 @@ def _sanitize_parameters(self, **kwargs):
...
+        candidate_labels = inputs["candidate_labels"]
+        if isinstance(candidate_labels, str):
+            candidate_labels = candidate_labels.split(",")
+
+        target_size = torch.IntTensor([[image.height, image.width]])
```
Comment (sgugger) [inline src/transformers/pipelines/zero_shot_object_detection.py:131]: I think PyTorch recommends using `torch.tensor` with the right `dtype`.

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py [hunk 4]
```diff
@@ -79,112 +79,96 @@ def test_small_model_pt(self):
...
         )
 
         self.assertEqual(
             nested_simplify(outputs, decimals=4),
+            # [
+            #     {"score": 0.7235, "label": "cat", "box": {"xmin": 204, "ymin": 167, "xmax": 232, "ymax": 190}},
+            #     {"score": 0.6748, "label": "remote", "box": {"xmin": 571, "ymin": 83, "xmax": 598, "ymax": 103}},
+            #     {"score": 0.6456, "label": "remote", "box": {"xmin": 494, "ymin": 105, "xmax": 521, "ymax": 127}},
+            #     {"score": 0.642, "label": "remote", "box": {"xmin": 67, "ymin": 274, "xmax": 93, "ymax": 297}},
+            # ],
```
Comment (sgugger) [inline tests/pipelines/test_pipelines_zero_shot_object_detection.py:93]: To clean up?

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py [hunk 5]
```diff
@@ -79,112 +79,96 @@ def test_small_model_pt(self):
...
         self.assertEqual(
             nested_simplify(outputs, decimals=4),
+            # [
+            #     [
+            #         {"score": 0.7235, "label": "cat", "box": {"xmin": 204, "ymin": 167, "xmax": 232, "ymax": 190}},
+            #         {"score": 0.6748, "label": "remote", "box": {"xmin": 571, "ymin": 83, "xmax": 598, "ymax": 103}},
+            #         {"score": 0.6456, "label": "remote", "box": {"xmin": 494, "ymin": 105, "xmax": 521, "ymax": 127}},
+            #         {"score": 0.642, "label": "remote", "box": {"xmin": 67, "ymin": 274, "xmax": 93, "ymax": 297}},
+            #     ]
+            # ],
```
Comment (sgugger) [inline tests/pipelines/test_pipelines_zero_shot_object_detection.py:126]: Same?

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! No problem with the breaking changes on my side since this a very new pipeline and thus still experimental.

File: src/transformers/pipelines/zero_shot_object_detection.py [hunk 6]
```diff
@@ -66,9 +80,7 @@ def __call__(
                 - A string containing a local path to an image
                 - An image loaded in PIL directly
 
-            text_queries (`str` or `List[str]` or `List[List[str]]`): Text queries to query the target image with.
-            If given multiple images, `text_queries` should be provided as a list of lists, where each nested list
-            contains the text queries for the corresponding image.
+            candidate_labels (`str` or `List[str]` or `List[List[str]]`): What the model should recognize in the image.
```
Comment (alaradirik) [inline src/transformers/pipelines/zero_shot_object_detection.py:83]: There is no need to input `List[List[str]]` `candidate_labels` if the input is always a single image, it'd also be better to expand the description :
```suggestion
            candidate_labels (`str` or `List[str]`): What the model should recognize in the image. If given a list, the input image is queried for each label / text query. 
```

File: src/transformers/pipelines/zero_shot_object_detection.py [hunk 7]
```diff
@@ -28,11 +20,33 @@
...
+    ...     "http://images.cocodataset.org/val2017/000000039769.jpg",
+    ...     candidate_labels=["cat", "couch"],
+    ... )
+    [{'score': 0.287, 'label': 'cat', 'box': {'xmin': 324, 'ymin': 20, 'xmax': 640, 'ymax': 373}}, {'score': 0.254, 'label': 'cat', 'box': {'xmin': 1, 'ymin': 55, 'xmax': 315, 'ymax': 472}}, {'score': 0.121, 'label': 'couch', 'box': {'xmin': 4, 'ymin': 0, 'xmax': 642, 'ymax': 476}}]
+
+    >>> detector(
+    ...         "https://huggingface.co/datasets/Narsil/image_dummy/raw/main/parrots.png",
+    ...         candidate_labels=["head", "bird"],
+    ... )
+    [{'score': 0.119, 'label': 'bird', 'box': {'xmin': 71, 'ymin': 170, 'xmax': 410, 'ymax': 508}}]
```
Comment (alaradirik) [inline src/transformers/pipelines/zero_shot_object_detection.py:44]: Could we replace this example with a more representative one?

Labels are not actually labels but free text queries ("a photo of a bird", "a close up view of xxx".  It'd also be less confusing to use an example where both queries are found.

**Review by alaradirik — APPROVED** [review]
Thanks for the PR! Switching to ChunkPipeline looks good to me.

As I mentioned in my previous comments, I have reservations about candidate_labels as it implies only one will be selected and that it should include labels rather than the free-form text ("a bird", " a photo of an orange bird", etc.) OWL-ViT is trained and evaluated on.

File: src/transformers/pipelines/zero_shot_object_detection.py (hunk 7 above)
Comment (Narsil) [reply src/transformers/pipelines/zero_shot_object_detection.py:44]: Then I am confused.

Isn't the pipeline supposed to return part of the image that correspond to the object in question ?
"A photo of  a bird" would be the entire image.
I tried to detect parts of the image other than `bird` but it always failed :(, that's why I used the cat image in the first place.

I do agree that finding an image where we could have different `candidate_labels` for the same image would be a much better test.

File: src/transformers/pipelines/zero_shot_object_detection.py (hunk 2 above)
Comment (Narsil) [reply src/transformers/pipelines/zero_shot_object_detection.py:126]: Done.

**Comment by Narsil** [comment]
I will wait a few days to give @sahamrit a chance to give his opinion before merging.

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py [hunk 8]
```diff
@@ -79,43 +79,32 @@ def test_small_model_pt(self):
...
             nested_simplify(outputs, decimals=4),
             [
-                [
-                    {"score": 0.7235, "label": "cat", "box": {"xmin": 204, "ymin": 167, "xmax": 232, "ymax": 190}},
-                    {"score": 0.6748, "label": "remote", "box": {"xmin": 571, "ymin": 83, "xmax": 598, "ymax": 103}},
-                    {"score": 0.6456, "label": "remote", "box": {"xmin": 494, "ymin": 105, "xmax": 521, "ymax": 127}},
-                    {"score": 0.642, "label": "remote", "box": {"xmin": 67, "ymin": 274, "xmax": 93, "ymax": 297}},
-                ]
+                {"score": 0.7235, "label": "cat", "box": {"xmin": 204, "ymin": 167, "xmax": 232, "ymax": 190}},
+                {"score": 0.7218, "label": "remote", "box": {"xmin": 204, "ymin": 167, "xmax": 232, "ymax": 190}},
```
Comment (sahamrit) [inline tests/pipelines/test_pipelines_zero_shot_object_detection.py:90]: @Narsil Something seems fishy here. First three have the exact same bounding box with different labels!

Comment (Narsil) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:90]: It's a random model, I think outputs shouldn't really matter (in terms of , is it correct output?).

Unless the model structure should disallow this, then yes something would be fishy.

Comment (sahamrit) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:90]: As per my understanding when Non-Max Suppression happens, the highest confidence box out of overlapping boxes are selected. I am not sure if this is done within a class or across classes. Since here the bounding boxes are exactly same, I doubt if there is something wrong with post processing. I also get your point of the model being random

Comment (Narsil) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:90]: Hmm now with ChunkPipeline all the labels are sent independantly.

Is the non max suppression done in the `feature_extractor` or in the model itself ?

File: src/transformers/pipelines/zero_shot_object_detection.py [hunk 9]
```diff
@@ -142,49 +140,54 @@ def _sanitize_parameters(self, **kwargs):
...
-            ]
-
-            result = sorted(result, key=lambda x: x["score"], reverse=True)
-            if top_k:
-                result = result[:top_k]
-            results.append(result)
+        for model_output in model_outputs:
+            label = model_output["candidate_label"]
+            model_output = BaseModelOutput(model_output)
+            outputs = self.feature_extractor.post_process(
```
Comment (sahamrit) [inline src/transformers/pipelines/zero_shot_object_detection.py:176]: @Narsil This is where I assume the non-max suppression would be done. We can check once though

Comment (Narsil) [reply src/transformers/pipelines/zero_shot_object_detection.py:176]: It was in the old code though: https://github.com/huggingface/transformers/pull/20280/files#diff-3ffc4527266bf533daeecb267c6e555a86fb8acd36c4207aecffb08c76b1f656L168

Comment (Narsil) [reply src/transformers/pipelines/zero_shot_object_detection.py:176]: Happy to switch though, this will need rework since I guess we need to stack all the outputs together before calliung this.

Is `post_process_image_guided_detection` defined for any other feature_extractor and/or is it easily implementable by any model that would loook like Owl ? (Just want to make sure this pipeline is not overfitting a single model)

File: tests/pipelines/test_pipelines_zero_shot_object_detection.py (hunk 8 above)
Comment (alaradirik) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:90]: @Narsil @sahamrit non-max suppression is not applied anywhere yet but I can add it to the feature extractor's post-process method once this PR is merged

Comment (sahamrit) [reply tests/pipelines/test_pipelines_zero_shot_object_detection.py:90]: In that case this PR LGTM!

File: src/transformers/pipelines/zero_shot_object_detection.py (hunk 9 above)
Comment (alaradirik) [reply src/transformers/pipelines/zero_shot_object_detection.py:176]: The old code only had the probability threshold but no NMS, could you add it back @Narsil? 

`post_process_image_guided_detection` is only implemented for OWL-ViT but it is applicable to all object detection post-process methods. I can open a separate PR to add NMS to the post-process methods of all object detection models, including OWL-ViT's zero-shot post-process method.

Comment (Narsil) [reply src/transformers/pipelines/zero_shot_object_detection.py:176]: If there's nothing specific about owl, We could also have the function directly within the pipeline. 

Or a util function that both the `feature_extractor` and the `pipeline could` use maybe ? :D.

Comment (alaradirik) [reply src/transformers/pipelines/zero_shot_object_detection.py:176]: It depends, supporting NMS in all post-processing methods would be really handy and I'll add it to all relevant models' feature extractors anyway. Or as you suggested, I can create a util function and edit the post-processing methods.

Comment (Narsil) [reply src/transformers/pipelines/zero_shot_object_detection.py:176]: Whichever your prefer. 

I like indedendant functions usually because there is less state to them (no `self.xxx` ) so it's easier to reason about what they do. Ofc this only works if the amount of arguments is limited.

Comment (alaradirik) [reply src/transformers/pipelines/zero_shot_object_detection.py:176]: Makes sense, I will go for the common util function for NMS + improved post-processing methods with `nms_threshold=0` for backward compatibility then.

Comment (Narsil) [reply src/transformers/pipelines/zero_shot_object_detection.py:176]: I will merge this and we can fix NMS in another PR, ok ?

Comment (alaradirik) [reply src/transformers/pipelines/zero_shot_object_detection.py:176]: Sounds good!
